### DISCUSSION TITLE
Authentication using custom secret

### DISCUSSION THREAD

**Discussion by yudiz-Manushi** [kiali/kiali#6940] [Q&A]
I've set up Kiali and implemented token authentication. However, it currently defaults to using the Kiali service account secret, whereas I would like to utilize a custom secret instead.
Kindly address this issue.
Istio verion : 1.15.0
kiali version : 1.55.*
Please find attached image.

![Screenshot 2023-12-08 at 3 48 24 PM](https://github.com/kiali/kiali/assets/152952067/3d7f2090-9652-4a04-9ab6-bda2a5839770)

**Comment by jmazzitelli** [comment]
(_moving to Q&A Discussion_)

**Comment by jmazzitelli** [comment]
Have you looked at the FAQ on this?

https://kiali.io/docs/faq/authentication/#how-to-obtain-a-token-when-logging-in-via-token-auth-strategy

For each user, they need their own secret and SA. They would then use the appropriate token. You must set up each user's SA according to their permissions.

**Reply by yudiz-Manushi** [reply]
@jmazzitelli I have refered mentioned document but I want to use a username/password authentication method.
So is there any method to do it??

**Reply by aljesusg** [reply]
Hi @yudiz-Manushi here you can check the authentication strategies supported by kiali

https://kiali.io/docs/configuration/authentication/

**Reply by yudiz-Manushi** [reply]
@aljesusg Yes I have checked mentioned document but username/password method is not defined.
So that's why I double check with support team.

**Reply by jmazzitelli** [reply]
If you want a user to authenticate using a username and password and access Kiali with those credentials (i.e."username/password method") there are three auth strategies Kiali supports that would allow for this.

You did not say what Kubernetes cluster you are using. If you are using OpenShift, then we recommend using the "[openshift](https://kiali.io/docs/configuration/authentication/openshift/)" auth strategy. When you use that, Kiali will utilize the OpenShift OAuth infrastructure and you log into Kiali using your OpenShift username/password.

If you are not using OpenShift, then a second auth strategy is "[openid](https://kiali.io/docs/configuration/authentication/openid/)" where you set up your own OpenID Connect infrastructure and configure Kiali to use it.

A third auth strategy is "[header](https://kiali.io/docs/configuration/authentication/header/)". This allows you to set up a reverse proxy in front of Kiali. This proxy can forward (via HTTP header) the user token that comes from an OAuth2 proxy or similar technology.

These three are explained in further detail in the doc that @aljesusg pointed to - https://kiali.io/docs/configuration/authentication/

**Comment by yudiz-Manushi** [comment]
Also I have another issue like kiali is not showing any data in UI.
Refer attached screenshots.
Let me know how do I resolve it?

![Screenshot 2023-12-11 at 1 06 54 PM](https://github.com/kiali/kiali/assets/152952067/979734f1-9568-4889-9062-178a0933f8fe)
![Screenshot 2023-12-11 at 1 10 29 PM](https://github.com/kiali/kiali/assets/152952067/ef76e4b6-19ad-4a50-8f10-49379e6eb6a0)

**Reply by aljesusg** [reply]
Hi @yudiz-Manushi we need more information to try to help you like the kiali configuration.

Anyway it seems something related with prometheus, can Kiali connect prometheus?

**Reply by yudiz-Manushi** [reply]
@aljesusg How do I Check if kiali connect to prometheus or not?

**Reply by aljesusg** [reply]
Check Kiali logs, prometheus pod try to make any query from kiali pod. We can't help you more without details 

You can see kiali config in the help button in the UI

### Environment
*Learn about how to determine versions [here](https://kiali.io/docs/faq/general/#how-do-i-determine-what-version-i-am-running).*

- **Kiali version:**
- **Istio version:**
- **Kubernetes impl:** 
- **Kubernetes version:** 
- **Other notable environmental factors:**

**Reply by yudiz-Manushi** [reply]
@aljesusg Please find attached details.
I have checked the kiali pod logs but did not find anything.
Kiali version: v1.55.1
Istio version: 1.15.0
Kubernetes impl: v1.27.7-eks-4f4795d
Kubernetes version: 1.27

**Reply by aljesusg** [reply]
@yudiz-Manushi Can you add the kiali configuration please? (You can find it in header/question button), anyway I am seeing a warning related with grafana unreachable. How did you install it?

Please add the output of **kubectl get svc -n istio-system** and **kubectl get pods -n istio-system**

**Reply by yudiz-Manushi** [reply]
@aljesusg I have installed grafana using helm chart.
All the svc and pods are running & up.
Please find attached screenshot.
Can we connect on google meet link if you are agree.
![Screenshot 2023-12-11 at 2 42 53 PM](https://github.com/kiali/kiali/assets/152952067/2f46a374-0e92-4154-a79a-4465f30439e0)

**Reply by yudiz-Manushi** [reply]
@aljesusg Just to add grafana is in different namespace.

**Reply by aljesusg** [reply]
wll...and prometheus? Prometheus store all the metrics from Istio...If you installed prometheus in another namespace too please add the kiali config because the point is taht kiali can't find prometheus like grafana and that's why you don't see the metrics

https://kiali.io/docs/configuration/p8s-jaeger-grafana/prometheus/
https://kiali.io/docs/configuration/p8s-jaeger-grafana/grafana/
https://kiali.io/docs/configuration/p8s-jaeger-grafana/tracing/

**Reply by yudiz-Manushi** [reply]
@aljesusg Yes grafana and prometheus both are in different namespace.
I have added URLs of grafana & prometheus in kiali config-map.
![Screenshot 2023-12-11 at 2 51 17 PM](https://github.com/kiali/kiali/assets/152952067/648c7ebb-a080-43d6-a4e2-32e97d38a500)

**Reply by aljesusg** [reply]
well this is not a tipical installation...check the kiali CR definition https://kiali.io/docs/configuration/kialis.kiali.io/#example-cr

I don't know if the prometheus  use auth or not, can you query from kiali pod to <prometehus-url>/api/v1/status/config ?

Kiali assumes a Prometheus server and is tested. So this kind of errors...

- Kiali may report that the TSDB is unreachable, and/or
- Kiali may show empty metrics if the TSBD does not implement the /api/v1/status/config

So please show the kiali logs because you should see and error message in logs when kiali pod start with some information about prometheus.

**Reply by yudiz-Manushi** [reply]
@aljesusg 
Can you please join meeting : To join the video meeting, click this link: https://meet.google.com/voe-drjh-zui
Otherwise, to join by phone, dial +1 620-869-1737 and enter this PIN: 340 858 267#
To view more phone numbers, click this link: https://tel.meet/voe-drjh-zui?hs=5

**Reply by yudiz-Manushi** [reply]
@aljesusg Also let me know how do I run this query from kiali pod?

**Reply by jmazzitelli** [reply]
There are some FAQs that you should review:

[Kiali does not work - What do i do?](https://kiali.io/docs/faq/general/#kiali-does-not-work---what-do-i-do) 
[Why is my graph empty?](https://kiali.io/docs/faq/graph/#emptygraph) 

And if you deploy Prometheus in a different namespace than istio-system control plane, you will need to tell Kiali that. See the [Kiali CR "external_services.prometheus" section docs here](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.prometheus).

> Can you please join meeting : To join the video meeting

This is a community project with no guaranteed level of support. It is a little much for a first time contributor to Kiali to expect a community member to immediately jump on a phone call to provide support.  If you wish to obtain personalized guaranteed product support for Kiali, please see the top of the [FAQ page](https://kiali.io/docs/faq/) : "_Need More Than Community Support? Enterprise support is [provided by Red Hat](https://www.redhat.com/en/technologies/cloud-computing/openshift/what-is-openshift-service-mesh)._"

**Reply by jmazzitelli** [reply]
BTW: I see you are new to github. May I explain also that you should create separate threads for different topics on github. It makes it easier for others to participate in the discussion threads (someone may know an answer to one issue, but not the other) and it makes searching the discussions for solutions  easier (someone with the same issue will search these threads and will also be able to use the solutions provided - mixing questions and answers in a single thread makes it harder to find information people are looking for).

You have three different questions here in this one discussion thread. (1) "How to use token auth with custom secrets/SAs",  (2) "How to enable username/passwords in Kiali" and (3) "How to configure Kiali when Prometheus is in another namespace".  We ask that you separate your questions like this into separate threads.
